contrib/epee/include/net/levin_protocol_handler_async.h

```diff
@@ -731,40 +734,65 @@ void async_protocol_handler_config<t_connection_context>::del_connection(async_p
 }
 //------------------------------------------------------------------------------------------
 template<class t_connection_context>
-void async_protocol_handler_config<t_connection_context>::del_out_connections(size_t count)
+void async_protocol_handler_config<t_connection_context>::delete_connections(size_t count, std::vector <boost::uuids::uuid> & connections)
 {
-	std::vector <boost::uuids::uuid> out_connections;
-	CRITICAL_REGION_BEGIN(m_connects_lock);
-	for (auto& c: m_connects)
-	{
-		if (!c.second->m_connection_context.m_is_income)
-			out_connections.push_back(c.first);
-	}
-	
-	if (out_connections.size() == 0)
-		return;
-
-	// close random out connections
+	// close random connections from  the provided set
 	// TODO or better just keep removing random elements (performance)
 	unsigned seed = std::chrono::system_clock::now().time_since_epoch().count();
-	shuffle(out_connections.begin(), out_connections.end(), std::default_random_engine(seed));
-	while (count > 0 && out_connections.size() > 0)
+	shuffle(connections.begin(), connections.end(), std::default_random_engine(seed));
+	while (count > 0 && connections.size() > 0)
 	{
 		try
 		{
-			auto i = out_connections.begin();
+			auto i = connections.begin();
 			async_protocol_handler<t_connection_context> *conn = m_connects.at(*i);
 			del_connection(conn);
 			close(*i);
-			out_connections.erase(i);
+			connections.erase(i);
 		}
 		catch (const std::out_of_range &e)
 		{
 			MWARNING("Connection not found in m_connects, continuing");
 		}
 		--count;
 	}
-	
+}
+//------------------------------------------------------------------------------------------
+template<class t_connection_context>
+void async_protocol_handler_config<t_connection_context>::del_out_connections(size_t count)
+{
+	std::vector <boost::uuids::uuid> out_connections;
+	CRITICAL_REGION_BEGIN(m_connects_lock);
+	for (auto& c: m_connects)
+	{
+		if (!c.second->m_connection_context.m_is_income)
+			out_connections.push_back(c.first);
+	}
+
+	if (out_connections.size() == 0)
+		return;
+
```

Comment:
That whole part can be factored too, since the only difference is a boolean which can be passed.

Reply:
Wouldn't t be better to change that to:
```
	if (out_connections.size() > 0)
		delete_connections(count, out_connections);
```
The early return without executing `CRITICAL_REGION_END` is a bit problematic.

Reply:
Nevermind. Now I get it. Will fix tomorrow.